File: src/octset.py

"""
The octset module implements heuristic algorithms for finding odd-cycle transversals (deletion set to bipartite).

Methods
-------
prescibed_octset(G, filename)
    Reads a predetermined octset for the graph G in from a file.
verify_bip(G, octset)
    Uses a BFS-coloring to find on octset for the graph G.  The octset can be given to the algorithm to determine its validity.
find_octset(G)
    Finds an octset on G by producing two maximal independent sets to make up the bipartite portion of the graph.  The remaining vertices are returned as the octset.
"""

import random

from collections import deque, OrderedDict

def prescibed_octset(graph, filename):
    """
    Reads a predetermined octset from a file and verifies its correctness.

    Parameters
    ----------
    graph : Graph
        The graph corresponding to the octset.  See graph.py.
    filename : str
        Name of the file containing the octset.

    Returns
    -------
    (set, set, set)
        A tuple of sets of size three corresponding to the octset, the left partition, and the right partition respectively.
    """

    octset = set()

    # opens the file
    with open(filename, "r") as f:
        for line in f:
            octset.add(int(line))

    # verifies that removing the octset makes the graph bipartite
    nu, left, right = verify_bip(graph, octset)
    return octset, left, right

def verify_bip(G, octset=set()):
    """
    Finds an octset on G by doing BFS.  When a vertex is added to both the left and the right, it gets added to the octset.

    Parameters
    ----------
    G : Graph
        The graph to solve oct on.  See graph.py.
    octset : set
        Optional.  Seeds the octset with vertices to add correctness-checking functionality.

    Returns
    -------
    (set, set, set)
        A tuple of sets of size three corresponding to the octset, left partition, and right partition respectively.
    """

    # sets up data structures for bfs
    queue = deque()
    unvisited = set(G.neighbors.keys() - octset)

    left = set()
    right = set()

    while unvisited or queue:
        # handles disconnected case
        if not queue:
            current = unvisited.pop()

            # adds first vertex to the left
            left.add(current)
        else:
            current = queue.popleft()

        for nbr in G.neighbors[current]:
            if nbr in octset:
                continue

            # adds each neighbor to the opposite side
            side = right if current in left else left
            side.add(nbr)

            # if the neighbor is now on both sides, must go in oct
            if nbr in left and nbr in right:
                left.remove(nbr)
                right.remove(nbr)
                octset.add(nbr)

            # add neighbor to queue if unseen, bfs
            if nbr in unvisited:
                unvisited.remove(nbr)
                queue.append(nbr)

    return octset, left, right

def find_octset(G):
    """
    Finds an octset on G by constructing two maximal independent sets in sequence.  Any vertices not in either independent set are added to the octset.

    Note that two implementations of maximal_indset are given.  The first (commented out) does not perform resorting based on degree and thus runs faster but produces larger octsets as a result.

    Parameters
    ----------
    G : Graph
        The graph to solve oct on.  See graph.py.

    Returns
    -------
    (set, set, set)
        A tuple of sets of size three corresponding to the octset, left partition, and right partition respectively.
    """

    # def maximal_indset(G, allowed):
    #     # choice between random selection and favoring small degree vertices
    #     # temp = sorted(allowed, key = lambda x: len(G.neighbors[x]))
    #     temp = random.sample(allowed, len(allowed))
    #
    #     # copy vertices to ordereddict for efficiency
    #     allowed = OrderedDict()
    #     for v in temp:
    #         allowed[v] = None
    #
    #     indset = set()
    #
    #     while allowed:
    #         # adds a vertex to the independent set
    #         (current, nu) = allowed.popitem()
    #         indset.add(current)
    #
    #         # removes all neighbors so they can't be added to the independent set
    #         for v in G.neighbors[current]:
    #             if v in allowed:
    #                 allowed.pop(v)
    #
    #     return indset

    def maximal_indset(G, allowed):
        deg = {}
        maxdeg = 0
        mindeg = len(allowed)

        # computes the degree of each allowed vertex
        for v in G.neighbors.keys():
            if v not in allowed:
                continue

            deg[v] = 0
            for u in G.neighbors[v]:
                if u in allowed:
                    deg[v] += 1

            maxdeg = max(maxdeg, deg[v])
            mindeg = min(mindeg, deg[v])

        # creates a map of degrees to sets of vertices with that degree
        revdeg = [None] * (maxdeg + 1)
        for i in range(maxdeg + 1):
            revdeg[i] = set()

        for v in deg.keys():
            revdeg[deg[v]].add(v)

        indset = set()

        while deg.keys():
            # finds the lowest degree vertex
            for i in range(mindeg, maxdeg + 1):
                if revdeg[i]:
                    mindeg = i
                    break

            # randomly picks a lowest degree vertex
            u = random.sample(revdeg[mindeg], 1)[0]
            revdeg[mindeg].remove(u)
            mindeg = max(mindeg - deg[u], 0)
            deg.pop(u)
            for v in G.neighbors[u]:
                if v not in deg:
                    continue

                # removes neighbors
                revdeg[deg[v]].remove(v)
                deg.pop(v)

                # updates degrees
                for w in G.neighbors[v]:
                    if w in deg:
                        revdeg[deg[w]].remove(w)
                        deg[w] -= 1
                        revdeg[deg[w]].add(w)

            indset.add(u)

        return indset

    # octset begins as entire graph
    octset = set(G.neighbors.keys())
    left = maximal_indset(G, octset)

    # remove vertices which are in the left partition
    octset.difference_update(left)
    right = maximal_indset(G, octset)

    # remove vertices which are in the right partition
    octset.difference_update(right)
    return octset, left, right

File: sr_apx/octset/octset.cpp

#include "octset.hpp"

#include <deque>
#include <iostream>
#include <fstream>

OctDecomp* verify_bip(Graph* graph, Set* os) {
	Set* left = new Set();
	Set* right = new Set();
	Set* octset = new Set();

	Set visited;
	for (Set::Iterator it = os->begin(); it != os->end(); ++it) {
		visited.insert(*it);
	}

	auto giter = graph->begin();

	std::deque<int> queue;
	int current;

	while (visited.size() < graph->size() || !queue.empty()) {
		if (queue.empty()) {
			while (visited.contains(*giter)) {
				++giter;
			}

			current = *giter;
			left->insert(current);
			visited.insert(current);
		}
		else {
			current = queue.front();
			queue.pop_front();
		}

		Set* neighbors = graph->neighbors(current);
		for (Set::Iterator it = neighbors->begin(); it != neighbors->end(); ++it) {
			int nbr = *it;
			if (os->contains(nbr) || octset->contains(nbr)) {
				continue;
			}

			if (left->contains(current)) {
				right->insert(nbr);
			}
			else {
				left->insert(nbr);
			}

			if (left->contains(nbr) && right->contains(nbr)) {
				left->erase(nbr);
				right->erase(nbr);
				octset->insert(nbr);
				continue;
			}

			if (!visited.contains(nbr)) {
				visited.insert(nbr);
				queue.push_back(nbr);
			}
		}
	}

	OctDecomp* ret = new OctDecomp();
	ret->left = left;
	ret->right = right;
	ret->octset = octset;

	return ret;
}

OctDecomp* prescribed_octset(Graph* graph, const char* filename) {
	std::ifstream f;
	f.open(filename, std::ios::in);

	Set* octset = new Set();

	char s[100];
	while (f.getline(s, 100)) {
		int u;
		sscanf(s, "%d", &u);
		octset->insert(u);
	}

	f.close();

	OctDecomp* ret = verify_bip(graph, octset);
	if (ret->octset->size() > 0) {
		printf("not a valid prescribed octset");
		return NULL;
	}

	ret->octset = octset;
	return ret;
}

void remove_vertex_oct(Graph* g, Map<int>& deg, Map<Set>& revdeg, int u) {
	for (auto nbr = g->neighbors(u)->begin(); nbr != g->neighbors(u)->end(); ++nbr) {
		if (!deg.contains(*nbr)) {
			continue;
		}

		int degree = deg[*nbr];
		revdeg[degree].erase(*nbr);
		deg[*nbr] = degree - 1;
		revdeg[degree - 1].insert(*nbr);
	}

	revdeg[deg[u]].erase(u);
	deg.erase(u);
}

Set* maximal_indset(Graph* graph, Set* available) {
	Map<int> deg;
	Map<Set> revdeg;

	for (Set::Iterator iu = available->begin(); iu != available->end(); ++iu) {
		int u = *iu;
		deg[u] = 0;
		for (Set::Iterator iv = graph->neighbors(u)->begin(); iv != graph->neighbors(u)->end(); ++iv) {
			int v = *iv;
			if (available->contains(v)) {
				deg[u] += 1;
			}
		}

		revdeg[deg[u]].insert(u);
	}

	Set* indset = new Set();

	while (deg.size() > 0) {
		int mindeg;
		for (int i = 0; i <= deg[*(deg.begin())]; ++i) {
			if (revdeg[i].size() > 0) {
				mindeg = i;
				break;
			}
		}

		int u = *(revdeg[mindeg].begin());
		for (Set::Iterator iv = graph->neighbors(u)->begin(); iv != graph->neighbors(u)->end(); ++iv) {
			int v = *iv;
			if (!deg.contains(v)) {
				continue;
			}

			remove_vertex_oct(graph, deg, revdeg, v);
		}

		revdeg[0].erase(u);
		deg.erase(u);

		indset->insert(u);
		available->erase(u);
	}

	return indset;
}

OctDecomp* find_octset(Graph* graph) {
	Set* octset = new Set();
	for (auto iu = graph->begin(); iu != graph->end(); ++iu) {
		int u = *iu;
		octset->insert(u);
	}

	Set* left = maximal_indset(graph, octset);
	Set* right = maximal_indset(graph, octset);

	OctDecomp* ret = new OctDecomp();
	ret->left = left;
	ret->right = right;
	ret->octset = octset;

	return ret;
}
